### PR TITLE
Add SystemInstruction::CreateAccountWithSeed

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 support for multiple accounts with the same authority exists, but no provision for 
 creating accounts with addresses derived from the authority exists.  it'd be really nice to define a space of addresses reachable for create_account() that don't require generating an ephemeral keypair.

 #### Summary of Changes
 add a system instruction that allows signers to create and assign in a subset of the
 pubkey address space as derived addresses

Fixes #

### CODE REVIEW

File: runtime/src/system_instruction_processor.rs [hunk 1]
```diff
@@ -1,13 +1,45 @@
...
+    program_id: &Pubkey,
+) -> Result<(), InstructionError> {
+    // from is the source of the derived address, the caller must have
+    //  signed, even if no lamports will be transferred
+    if from.signer_key().is_none() {
+        debug!("CreateAccountWithSeed: from must sign");
+        return Err(InstructionError::MissingRequiredSignature);
+    }
+
+    // re-derive the address, must match to
```
Comment (mvines) [inline runtime/src/system_instruction_processor.rs]:
```suggestion
    // re-derive the address, must match `to`
```

File: runtime/src/system_instruction_processor.rs [hunk 2]
```diff
@@ -1,13 +1,45 @@
...
+
+fn create_account_with_seed(
+    from: &mut KeyedAccount,
+    to: &mut KeyedAccount,
+    seed: &str,
+    lamports: u64,
+    space: u64,
+    program_id: &Pubkey,
+) -> Result<(), InstructionError> {
+    // from is the source of the derived address, the caller must have
```
Comment (mvines) [inline runtime/src/system_instruction_processor.rs]:
```suggestion
    // `from` is the source of the derived address, the caller must have
```

File: sdk/src/system_instruction.rs [hunk 3]
```diff
@@ -38,6 +45,20 @@ pub enum SystemInstruction {
...
     },
+    /// Create a new account at an address derived from
+    ///    the from account and a seed
+    /// * Transaction::keys[0] - source
+    /// * Transaction::keys[1] - new account key
+    /// * seed - string of ascii chars, no longer than MAX_ADDRESS_SEED_LEN
+    /// * lamports - number of lamports to transfer to the new account
+    /// * space - memory to allocate if greater then zero
+    /// * program_id - the program id of the new account
+    CreateAccountWithSeed {
```
Comment (mvines) [inline sdk/src/system_instruction.rs]: We'll need to adjust solana-web3.js once this lands, do you want a hand with that @rob-solana?

File: runtime/src/system_instruction_processor.rs [hunk 4]
```diff
@@ -1,13 +1,45 @@
...
+    instruction_processor_utils::{limited_deserialize, next_keyed_account},
+    pubkey::Pubkey,
+    system_instruction::{create_address_with_seed, SystemError, SystemInstruction},
+    system_program, sysvar,
+};
+
+fn create_account_with_seed(
+    from: &mut KeyedAccount,
+    to: &mut KeyedAccount,
+    seed: &str,
```
Comment (t-nelson) [inline runtime/src/system_instruction_processor.rs]: Any reason for the restriction to 32B ASCII string? `[u8; 32]` could support some internationalization and hash digests

Comment (mvines) [reply runtime/src/system_instruction_processor.rs]: yeah good point, we don't care that the seed is valid utf-8

**Comment by garious** [comment]
The implications here might be more far-reaching than intended. Without derived addresses, we can assert all account addresses are ed25519 pubkeys. And with that invariant we can say every account creator at one time possessed its private key. Is that a valuable invariant? Hard to say. BIP32 would preserve it, but not exactly an option for ed25519.

File: sdk/src/system_instruction.rs (hunk 3 above)
Comment (rob-solana) [reply sdk/src/system_instruction.rs]: lemme take a crack first

File: runtime/src/system_instruction_processor.rs (hunk 4 above)
Comment (rob-solana) [reply runtime/src/system_instruction_processor.rs]: I wanted to keep the space smaller than 256 bytes of seed to deter grinding attacks.  32 ascii chars takes away 32 bits.

**Comment by rob-solana** [comment]
> The implications here might be more far-reaching than intended. Without derived addresses, we can assert all account addresses are ed25519 pubkeys. And with that invariant we can say every account creator at one time possessed its private key. Is that a valuable invariant? Hard to say. BIP32 would preserve it, but not exactly an option for ed25519.

does Pubkey::new_rand() (a perfectly acceptable `to` for a transfer) guarantee an ed25519 pubkey?

File: runtime/src/system_instruction_processor.rs (hunk 4 above)
Comment (aeyakovenko) [reply runtime/src/system_instruction_processor.rs]: @rob-solana grinding on what though?

Comment (rob-solana) [reply runtime/src/system_instruction_processor.rs]: trying to DoS a guy you think wants pubkey+"my secret stake \#23"

Comment (rob-solana) [reply runtime/src/system_instruction_processor.rs]: if we think 256 bits of seed is ok, I'll just accept another Hash instead of an ascii string.

Comment (aeyakovenko) [reply runtime/src/system_instruction_processor.rs]: @rob-solana you can always grind with pubkey generation.  I think we should either have a Hash/[u8;32], or u64.

**Comment by garious** [comment]
> does Pubkey::new_rand() (a perfectly acceptable to for a transfer) guarantee an ed25519 pubkey?

I've always been quietly uncomfortable with `new_rand` since I wrote it, but didn't want the ed25519 dependency in pubkey.rs, which needs to compile to BPF. In any case, the unit-testing utility doesn't set a precedent. It can change. So can the genesis accounts. This instruction, however, once added to the System interface, is forever.

**Comment by aeyakovenko** [comment]
@garious there is no way to prevent a transfer to a random address.  Destinations don’t require a proof.

File: runtime/src/system_instruction_processor.rs (hunk 4 above)
Comment (t-nelson) [reply runtime/src/system_instruction_processor.rs]: Requiring `from`'s signature effectively makes griefing the address space a second-preimage attack on SHA256.  We should probably be OK with that.  Though fewer bits _do_ minimize opportunities for shenanigans.  Would it make sense to restrict the seed to something as small as a `u16`?

**Comment by garious** [comment]
> there is no way to prevent a transfer to a random address. Destinations don’t require a proof.

There may be a way to enforce the address is a valid pubkey. We should consider locking it down and replacing new_rand() with

**Comment by garious** [comment]
I'm not saying we should go one way or another on this, but it's generally not a great idea to define a data type as the bits of either one data type or another, depending on some usage (semantic coupling). Effectively, that makes Pubkey a C-style union representing either a public key or a random string of bits. You might argue, no, it's just a random string of bits, but in fact we interpret it as a public key quite often (i.e. System::Transfer).  And if random bits, why are we using Pubkey? I'd suggest sleeping on it. There's probably a more elegant solution here - maybe defining a new type AccountAddress as an enum instead of a Pubkey.

File: runtime/src/system_instruction_processor.rs [hunk 5]
```diff
@@ -50,11 +92,18 @@ fn create_system_account(
         return Err(SystemError::ResultWithNegativeLamports.into());
     }
 
-    assign_account_to_program(to, program_id)?;
```
Comment (ryoqun) [inline runtime/src/system_instruction_processor.rs]: Not calling `assign_account_to_program` anymore for normal accounts starts to skip [this existing check for account.signer_key](https://github.com/solana-labs/solana/pull/7390/files#diff-f06d4cdb8c607b22d0504416e326b5cfL65). Is that OK?

Comment (rob-solana) [reply runtime/src/system_instruction_processor.rs]: that check is already present in `create_account()` and is not enforced by `create_account_with_seed()`

this change was intentional

**Comment by rob-solana** [comment]
> I'm not saying we should go one way or another on this, but it's generally not a great idea to define a data type as the bits of either one data type or another, depending on some usage (semantic coupling). Effectively, that makes Pubkey a C-style union representing either a public key or a random string of bits. You might argue, no, it's just a random string of bits, but in fact we interpret it as a public key quite often (i.e. System::Transfer). And if random bits, why are we using Pubkey? I'd suggest sleeping on it. There's probably a more elegant solution here - maybe defining a new type AccountAddress as an enum instead of a Pubkey.

sleeping on it good.

I wonder if we can force the resulting generated addresses to be *unreachable* pubkeys in ed25519

File: runtime/src/system_instruction_processor.rs (hunk 5 above)
Comment (ryoqun) [reply runtime/src/system_instruction_processor.rs]: Ok, thanks for clarifying!

**Comment by aeyakovenko** [comment]
@rob-solana @garious we can set a bit in the Account to indicate that its not a real pubkey and cannot be signed with.  that makes more sense than changing the address type.

**Comment by garious** [comment]
Perhaps the `Account.owner` could be the dynamic tag that implies how an `AccountAddress` should be interpreted.  Or perhaps the "seed" needs to be an existing account address. Just some thoughts...

**Comment by rob-solana** [comment]
the target customer for this work found `keypair::new() + create_account() + set_staker()` to be untenable (i.e. a burner keypair for the "derived" account), objected to the existence of the ephemeral keypair to sign the create_account() TX

**Comment by rob-solana** [comment]
> > there is no way to prevent a transfer to a random address. Destinations don’t require a proof.
> 
> There may be a way to enforce the address is a valid pubkey. We should consider locking it down and replacing new_rand() with Keypair::new().pubkey()

I think just recasting Pubkey::new_rand() to do that would work?

**Comment by aeyakovenko** [comment]
@garious seed as an existing account address won’t prevent a easy to factor pubkey.  I don’t see a problem of relaxing the account address = pubkey relationship.  We have no way to force addresses to be strong private keys.

**Comment by rob-solana** [comment]
> @garious seed as an existing account address won’t prevent a easy to factor pubkey. I don’t see a problem of relaxing the account address = pubkey relationship. We have no way to force addresses to be strong public keys.

In this discussion we should keep in mind that builtin `program_ids` and `sysvars` have this "not a `Pubkey` of a `Keypair`" property, too.

File: sdk/src/system_instruction.rs (hunk 3 above)
Comment (rob-solana) [reply sdk/src/system_instruction.rs]: ok, I failed: https://github.com/solana-labs/solana-web3.js/pull/571

File: runtime/src/system_instruction_processor.rs (hunk 4 above)
Comment (rob-solana) [reply runtime/src/system_instruction_processor.rs]: were it as small as a u16, mnemonics get harder.   ascii is nice for mnemonics (in English).  opening to utf-8 would make it look less amateur-hour.

The grinding attack is picking another source `from` (obs having the keypair), and griefing some other dude's `from`.  The goal is to make that griefing look hard or futile.  

I think we can do this with:
1. u64, maybe low 8 bytes of hash(mnemonic string)?
2. hashing a limited-length string as proposed.  Upgrading to utf-8 hurts the deterrent a little as you can cover 256 bits of space with only 13-ish utf-8 chars (at 20bits/char).

We can also say "who cares?  grind away!"

**Comment by garious** [comment]
I think our best path forward is a new 32 byte AccountAddress type with a From instance for Pubkey.

**Comment by t-nelson** [comment]
> maybe defining a new type AccountAddress as an enum instead of a Pubkey.

Bonus points for a potential curve migration mechanism

File: runtime/src/system_instruction_processor.rs (hunk 4 above)
Comment (rob-solana) [reply runtime/src/system_instruction_processor.rs]: I'm planning to remove the ascii stuff, appreciate input here @t-nelson @aeyakovenko @garious

Comment (t-nelson) [reply runtime/src/system_instruction_processor.rs]: > were it as small as a u16, mnemonics get harder. ascii is nice for mnemonics (in English). opening to utf-8 would make it look less amateur-hour.

Right, the `u16` suggestion would discard the mnemonic for an index (`from` gets `I` accounts under `program_id`). Which gives cheap introspection, but no pseudonymity.

> The grinding attack is picking another source `from` (obs having the keypair), and griefing some other dude's `from`. The goal is to make that griefing look hard or futile.

Yep. A SHA256 second preimage search is pretty futile IMO.  Restricting free bits per `from` is about the most effective deterrent we can enact.  Grinding `from`s is always possible and slower than the hashing step.  You've already covered the other speedup.  That is, requiring fixed bits at the end of the input data, preventing working from a midstate.
  
> I think we can do this with:
> 
>     1. u64, maybe low 8 bytes of hash(mnemonic string)?
> 
>     2. hashing a limited-length string as proposed.  Upgrading to utf-8 hurts the deterrent a little as you can cover 256 bits of space with only 13-ish utf-8 chars (at 20bits/char).

I think (valid) utf-8 would actually be _more_ restricted.  For non-ASCII code points, the top *two* bits of every byte are fixed and the first byte also consumes up to three more bits to encode the length.  Not all code pages are in use or full either, further restricting valid bitstrings

> We can also say "who cares? grind away!"

Probably fine. See, https://lbc.cryptoguru.org/about

**Comment by rob-solana** [comment]
updated to use a 32-char utf-8 string, which is equivalently or possibly more restrictive in grinding from seed as ascii

**Pull request merged**